### DISCUSSION TITLE
AbstractClass에서 python의 abc 사용하기

### DISCUSSION THREAD

**Discussion by anohk** [8percent/styleguide#28] [General]
추상 클래스를 작성할 때, python의 abc를 사용하는 것을 제안합니다.

https://www.python.org/dev/peps/pep-3119/
https://docs.python.org/3/library/abc.html#module-abc

현재는 구체 클래스에서 필수적으로 구현해야하는 메서드나 프로퍼티가 있는 경우,
추상클래스에서 NotImplementedError 를 발생시키는 방식으로 작성되고 있습니다.

python의 abc를 사용하게 되면 상속구조에 대해 조금 더 strict한 방식으로 관리할 수 있는 장점이 있습니다.

## 기존에 사용하던 방식
```python
class AbstractClass:

    def foo(self):
        return NotImplementedError

    @property
    def bar(self):
        return NotImplementedError
````
- 위와 같은 경우 AbstractClass는 추상클래스로 제공할 의도이지만 인스턴스화 할 수 있습니다.
- 파생 클래스에서 메서드 구현이 되어 있지 않은 경우, 해당 메서드 호출 시점에 NotImplementedError가 발생합니다.

## abc 사용
```python
class AbstractClass(ABC):

    @abstractmethod
    def foo(self):
        pass

    @property
    @abstractmethod
    def bar(self):
        pass
```
- abc를 사용하는 경우, 추상 클래스를 인스턴스화 할 수 없습니다.
- `@abstractmethod` 데코레이터를 적용하여 파생 클래스에서 반드시 구현해야하는 메서드를 강제할 수 있습니다.
- 파생 클래스에서 메서드 구현이 되어 있지 않은 경우, 인스턴스를 생성하는 시점에 에러가 발생합니다.

----------

### 추가) 추상 모델 클래스에도 abc를 적용해 볼 수 있는가?

추상 모델 클래스의 경우 ABC를 적용하면 metaclass conflict 가 발생합니다. 

```python
class AbstractModel(models.Model):
     class Meta:
         abstract = True

class AbstractClass(ABC):
     @abstractmethod
     def foo(self):
         pass

class DerivedClass(AbstractModel, AbstractClass):
     def foo(self):
         return
```

```
metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

아래와 같이 meta class를 재정의해서 상속하는 방식을 사용하는 것이 해결책으로 보입니다.
```python
class AbstractModelMeta(ABCMeta, type(models.Model)):
    ...

class AbstractClass(models.Model, metaclass=AbstractModelMeta):
    class Meta:
        abstract = True

    @abstractmethod
    def foo(self):
        pass

class DerivedClass(AbstractClass):
    ...
```

추상 모델 클래스의 경우 abc 사용하는 것이 적절할 지는 고민이 되는데 다른 분들의 의견이 궁금합니다.

**Comment by Kirade** [comment]
ABC 를 사용하는 경우 추상 클래스를 상속하는 구체 클래스를 구현할 때 발생할 수 있는 실수를 줄일 수 있다는 점에서,
필요한 경우 사용할 수 있을 것 같습니다.

장고 모델의 경우 ABC 를 사용하는것은 적절하지 않다고 생각합니다.
장고에서 추상 모델을 정의할 때 제안하는 방식이 아니고, 모델의 경우 메타클래스를 별도로 정의하는데 이것과 ABCMeta 가 서로 충돌할 가능성도 있을것 같습니다.

**Comment by donghyun0218** [comment]
저도 ABC 를 사용하는 것에 동의합니다.

**Comment by kyungjunleeme** [comment]
저는 장고 모델의 경우 ABC 를 사용하는 것이 적절하지 않다고 생각합니다. ABC를 사용하는 것에 대한 내용을 공식문서에도 찾아볼 수 없고,

two scoop of django 3.x 에서 발췌된 내용에 근거해서 좋지 않다고 생각합니다. 
``` 
Don’t confuse Django abstract base classes with the abstract base classes in the Python standard library’s abc module, as they have very different purposes and behaviors.
```

**Comment by leehosung** [comment]
저도 같은 의견입니다. 일반 클래스는 ABC를 권고, Django 추상 클래스는 사용하지 않는것으로.

**Comment by cbscsm** [comment]
저도 장고 모델 이외 클래스에 대해서는 ABC 사용에 동의합니다. 특히 Mixin 같은 경우 ABC를 사용하면 정적분석 커버리지를 크게 향상시킬 수 있습니다.
또한 재정의가 필수로 요구되는 메서드에 대해서도 공통 부분을 구현(super를 통해 호출 가능)할 수 있다는 점이, NotImplementError를 발생 시키는 것 보다 더 좋다고 생각합니다.
장고 모델에서는 저도 다른분들 의견에 동의합니다. 하위 클래스에서 메타 클래스를 지정하면, 기존 상위클래스의 메타 클래스가 쓰이지 않게 되기 때문에 동작에서 문제가 생길 여지가 있을 것 같습니다.
